### PR TITLE
[CARBONDATA-1288][DictionarySecureServer] Dictionary Secure Server Implementation

### PR DESCRIPTION
**Pull request by sounakr**
During single pass load  Dictionary key creation is done through driver and executor communication through external ports. The communication will happen through TCP communication which is not secure or encrypted.
But in case spark turn on Security parameters then carbon dictionary ports also has to follows same authentication and encryption i.e. communicate through SASL Authentication protocol and Digest-MD5 encryption.
This PR makes the dictionary server and client communication Secure and encrypted. In case spark turn ON security and authentication through the below parameters then Carbon communication also becomes secure. 
By default the communication is still non secure. 

Parameters 
**spark.authenticate true
spark.authenticate.enableSaslEncryption true
spark.authenticate.secret** 

 - [X] Any interfaces changed?
 No
 - [X] Any backward compatibility impacted?
 No
 - [X] Document update required?
Yes
 - [X] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [X] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/751/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1981/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2301/

**Comment by sounakr** [comment]
Retest this please.

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1992/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/763/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2307/

**Comment by sounakr** [comment]
Retest this please.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1996/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/766/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2310/

**Comment by sounakr** [comment]
Retest this please

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/770/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2002/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2345/

File: core/src/main/java/org/apache/carbondata/core/dictionary/server/NonSecureDictionaryServer.java [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    this.port = port;
+    startServer();
+  }
+
+  public static synchronized DictionaryServer getInstance(int port, CarbonTable carbonTable)
+      throws Exception {
+    if (INSTANCE == null) {
+      INSTANCE = new NonSecureDictionaryServer(port);
+    }
+    INSTANCE.initializeDictionaryGenerator(carbonTable);
```
[the remark this answers is not in the document]
Comment (sounakr) [reply core/src/main/java/org/apache/carbondata/core/dictionary/server/NonSecureDictionaryServer.java]: Yes, Initially the code was like 
   if (INSTANCE == null) {
       synchronized(lock) {
          if (INSTANCE == null) 
            ...
            ... 
There was a dual check of INSTANCE for concurrency. The dual check will cause a find bug error. In order to void this the method is turned into synchronized method. Also initializeDictionaryGenerator will finally invoke method like initializeGeneratorForTable which is also synchronized. So even though we remove the method as synchronized the execution will never be concurrent.

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/804/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2035/

File: core/src/main/java/org/apache/carbondata/core/dictionary/service/NonSecureDictionaryServiceProvider.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.carbondata.core.dictionary.service;
+
+import org.apache.carbondata.core.dictionary.client.DictionaryClient;
+import org.apache.carbondata.core.dictionary.client.NonSecureDictionaryClient;
+
+public class NonSecureDictionaryServiceProvider implements DictionaryServiceProvider {
+  private int port = 0;
```
[the remark this answers is not in the document]
Comment (sounakr) [reply core/src/main/java/org/apache/carbondata/core/dictionary/service/NonSecureDictionaryServiceProvider.java]: Done.

**Comment by sounakr** [comment]
Retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/816/

**Comment by sounakr** [comment]
Retest this please

**Comment by sounakr** [comment]
Retest this please

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/845/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2070/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2373/

**Comment by sounakr** [comment]
Retest this please.

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2072/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/846/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2374/

**Comment by sounakr** [comment]
Retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/847/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2073/

**Comment by gvramana** [comment]
LGTM

**Pull request closed without merging**